salt/_modules/caasp_etcd.py

```diff
@@ -19,37 +42,42 @@ def get_cluster_size():
     less than 3, it will bump it to 3 (or the number of nodes
     available if the number of nodes is less than 3).
     """
-    member_count = __pillar__["etcd"]["masters"]
+    member_count = __salt__['pillar.get']('etcd:masters', None)
 
-    if member_count is not None:
-        # A value has been set in the pillar, respect the users choice
-        # even it's not a "good" number.
-        member_count = int(member_count)
+    if not member_count:
+        # A value has not been set in the pillar, calculate a "good" number
+        # for the user.
+        num_masters = _get_num_kube("roles:kube-master")
 
-        if member_count < DESIRED_MEMBER_COUNT:
-            LOG.warning("etcd member count too low (%d), consider increasing "
-                        "to %d", member_count, DESIRED_MEMBER_COUNT)
+        # considering some intervals, find the optimal number of etcd members
+        if num_masters >= 7:
+            member_count = 7
+        elif num_masters >= 5:
+            member_count = 5
+        elif num_masters >= RECOMMENDED_MEMBER_COUNT:
+            member_count = RECOMMENDED_MEMBER_COUNT
+        elif num_masters < RECOMMENDED_MEMBER_COUNT:
+            # Attempt to increase the number of etcd master to 3,
+            # however, if we don't have 3 nodes in total,
+            # then match the number of nodes we have.
+            increased_member_count = _get_num_kube("roles:kube-*")
+            increased_member_count = min(
+                RECOMMENDED_MEMBER_COUNT, increased_member_count)
 
-            return member_count
+            member_count = increased_member_count
```

Comment:
I think we should ensure `member_count` is an odd number to ensure etcd leader election can be done. I would decrease `member_count` by one if needed

Reply:
Ok, we can do that...